### PR TITLE
[azure] Updated coverage report

### PR DESCRIPTION
**Pull request by michelececcacci**

### CODE REVIEW

File: scripts/templates/generate_coverage_report [hunk 1]
```diff
@@ -125,24 +129,38 @@ if [[ "$1" == "--codecov" ]]; then
       echo "WARNING: No CODECOV_TOKEN defined. Upload to CodeCov might fail"
    fi
 
-   # Remove other files to be sure CodeCov uses only the coverage.info file
-   rm -f lcov.info
-   remove_gcda_files
-
    # Upload report to CodeCov
+   if [ ! -d build/tmp ]; then
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:133]: Style: `if ! [ -d ....`

File: scripts/templates/generate_coverage_report [hunk 2]
```diff
@@ -112,6 +112,10 @@ if [[ "$1" == "--codecov" ]]; then
       exit 1
    fi
 
+   # Remove other files to be sure CodeCov uses only the coverage.info file
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:115]: Unnecessary change: these lines were blow the TOKEN check. Please, don't move them around or the diff gets tricker than necessary to review.

File: scripts/templates/generate_coverage_report (hunk 1 above)
Comment (vvaltchev) [reply scripts/templates/generate_coverage_report:133]: **BUG**: Don't mention the "build" directory like that: there is the `$BUILD` variable specifically for that. See above.
The build directory doesn't have to be `TILCK_ROOT/build`

File: scripts/templates/generate_coverage_report [hunk 3]
```diff
@@ -125,24 +129,38 @@ if [[ "$1" == "--codecov" ]]; then
       echo "WARNING: No CODECOV_TOKEN defined. Upload to CodeCov might fail"
    fi
 
-   # Remove other files to be sure CodeCov uses only the coverage.info file
-   rm -f lcov.info
-   remove_gcda_files
-
    # Upload report to CodeCov
+   if [ ! -d build/tmp ]; then
+      mkdir build
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:134]: Same as above. Use mkdir -p "$BUILD/tmp"

File: scripts/templates/generate_coverage_report [hunk 4]
```diff
@@ -125,24 +129,38 @@ if [[ "$1" == "--codecov" ]]; then
...
       fi
+
    fi
 
+   chmod +x codecov
+
+   if [ ! -z  $"CODECOV-TOKEN" ]; then
+      ./codecov
+   else
+      ./codecov -t $"CODECOV-TOKEN"
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:162]: **BUG:** First, you don't need to do that at all. If the env variable has been exported, the codecov tool will see it.
Also, the name was `CODECOV_TOKEN`. Third, $"something" doesn't mean anything to me. "$something" is the correct form.

File: scripts/templates/generate_coverage_report [hunk 5]
```diff
@@ -125,24 +129,38 @@ if [[ "$1" == "--codecov" ]]; then
...
-   remove_gcda_files
-
    # Upload report to CodeCov
+   if [ ! -d build/tmp ]; then
+      mkdir build
+      mkdir build/tmp
+   fi
 
-   if ! bash <(curl -s https://codecov.io/bash) -X gcov; then
+   cd build/tmp/
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:138]: **BUG:** Don't change the directory like that. It might not work.
Do `pushd $BUILD/tmp` and later, after the chmod, do `popd`, to the restore the CWD.
Also, later run codecov from the current directory with: `$BUILD/tmp/codecov`.

**Review by vvaltchev — CHANGES_REQUESTED** [review]

File: scripts/templates/generate_coverage_report (hunk 4 above)
Comment (vvaltchev) [reply scripts/templates/generate_coverage_report:162]: **BUG:** the -X gcov option is missing.

Comment (michelececcacci) [reply scripts/templates/generate_coverage_report:162]: So i just need to do `$BUILD/tmp/codecov -X gcov`?

File: scripts/templates/generate_coverage_report (hunk 3 above)
Comment (michelececcacci) [reply scripts/templates/generate_coverage_report]: Shouldn't it be "$BUILD"/tmp?

Comment (vvaltchev) [reply scripts/templates/generate_coverage_report]: Absolutely not. Scripting languages like Bash, Perl, PHP and many others (in particular shells), have been designed to expand variables inside strings. In case the next char matches [A-Za-z0-9_], then the correct syntax would be "${BUILD}something". The weird syntax `"$BUILD"/tmp` might work on Bash, but it's generally unportable.

File: scripts/templates/generate_coverage_report (hunk 4 above)
Comment (vvaltchev) [reply scripts/templates/generate_coverage_report]: Yes, of course. The whole point was to replace `bash <(curl -s https://codecov.io/bash)` with `$BUILD/tmp/codecov` everywhere after downloading it.

File: scripts/templates/generate_coverage_report (hunk 3 above)
Comment (michelececcacci) [reply scripts/templates/generate_coverage_report]: Ok, thanks for the answer!

File: scripts/templates/generate_coverage_report (hunk 4 above)
Comment (michelececcacci) [reply scripts/templates/generate_coverage_report]: Perfect!

File: scripts/templates/generate_coverage_report [hunk 6]
```diff
@@ -112,6 +112,7 @@ if [[ "$1" == "--codecov" ]]; then
       exit 1
    fi
 
+
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:115]: Extra whitespace, please remove.
Nit: after making a change and saving the commit, just run: `git diff HEAD~1` to inspect your change` and make sure doesn't have any unnecessary things :-)

File: scripts/templates/generate_coverage_report [hunk 7]
```diff
@@ -130,19 +131,35 @@ if [[ "$1" == "--codecov" ]]; then
...
+      curl -Os https://uploader.codecov.io/latest/linux/codecov
+      if ! [ -f codecov ]; then
+         echo "ERROR: codecov did not collect coverage reports"
+         echo "Trying again in 1 minute..."
+         sleep 60
+      fi
 
-      if ! bash <(curl -s https://codecov.io/bash) -X gcov; then
+      curl -Os https://uploader.codecov.io/latest/linux/codecov
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:149]: **BUG:** this will run unconditionally, even if the first curl succeeded!
There was a reason why that was under an `if` in the old code.
Please fix it.

File: scripts/templates/generate_coverage_report [hunk 8]
```diff
@@ -130,19 +131,35 @@ if [[ "$1" == "--codecov" ]]; then
...
+      if ! [  -f codecov ]; then
          echo "FATAL ERROR: Codecov cannot collect the coverage reports"
          exit 1
       fi
+
    fi
 
+   chmod +x codecov
+
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:158]: style: remove this extra empty line

File: scripts/templates/generate_coverage_report [hunk 9]
```diff
@@ -130,19 +131,35 @@ if [[ "$1" == "--codecov" ]]; then
...
+   popd
+
+   "$BUILD/tmp/codecov" -X gcov
+
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:162]: style: remove this extra empty line

File: scripts/templates/generate_coverage_report [hunk 10]
```diff
@@ -130,19 +131,35 @@ if [[ "$1" == "--codecov" ]]; then
...
          echo "FATAL ERROR: Codecov cannot collect the coverage reports"
          exit 1
       fi
+
    fi
 
+   chmod +x codecov
+
+   popd
+
```
Comment (vvaltchev) [inline scripts/templates/generate_coverage_report:160]: style: this empty line must be kept, instead.
```
popd
chmod
<empty line>
codecov
exit
```

**Review by vvaltchev — CHANGES_REQUESTED** [review]

**Review by vvaltchev — COMMENTED** [review]
Great, thanks for fixing!
